lib/sexp.nit

```diff
@@ -0,0 +1,170 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# This file is free software, which comes along with NIT.  This software is
+# distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without  even  the implied warranty of  MERCHANTABILITY or  FITNESS FOR A
+# PARTICULAR PURPOSE.  You can modify it is you want,  provided this header
+# is kept unaltered, and a notification of the changes is added.
+# You  are  allowed  to  redistribute it and sell it, alone or is a part of
+# another product.
+
+# S-Expression parsing facilities
+module sexp
+
+intrude import parser_base
+
+# Any S-Expression entity
+abstract class SExpEntity
+
+	# Location in the source document
+	var location: nullable Location
+end
+
+# A full S-Expression, delimited by `(` and `)`
+class SExp
+	super SExpEntity
+
+	# Children of a SExp
+	var content = new Array[SExpEntity]
+
+	redef fun to_s do return "({content.join(" ")})"
+
+	# Returns a pretty-printable version of self
+	fun pretty_to_s: String do return recurse_to_s(0)
+
+	private fun recurse_to_s(depth: Int): String do
+		var s = "{"\t" * depth}(\n"
+		for i in content do
+			if i isa SExp then
+				s += i.recurse_to_s(depth + 1)
+				s += "\n"
+				continue
+			end
+			s += "\t" * (depth + 1)
+			s += i.to_s
+			s += "\n"
+		end
+		return s + "{"\t" * depth})"
+	end
+end
+
+# A Double-quoted String
+class SExpDQString
+	super SExpEntity
+
+	# Double-quoted string
+	var content: String
+
+	redef fun to_s do return content
+end
+
+# A float-value
+class SExpFloat
+	super SExpEntity
+
+	# Floating-point value
+	var content: Float
+
+	redef fun to_s do return content.to_precision(2)
+end
+
+# Any Identifier, non string and non-float
+class SExpId
+	super SExpEntity
+
+	# S-Exp compatible identifier
+	var content: String
+
+	redef fun to_s do return content
+end
+
+# An error parsing S-Expressions
+class SExpError
+	super SExpEntity
+
+	# Error message
+	var message: String
+
+	redef fun to_s do return "S-Expression error: {message} at {location or else "unknown location"}"
+end
+
+# S-Expression processor
+class SExpProcessor
+	super StringProcessor
+
+	# Parses an S-Expression
+	fun parse_document: SExpEntity do
+		var stack = new Array[SExp]
+		var srclen = src.length
+		ignore_whitespaces
+		if pos >= srclen then return new SExpError("Empty S-expression")
+		var c = src[pos]
+		if c != '(' then return new SExpError("Invalid start of S-Expression, got `{c}`, exprected `(`.")
```

Comment:
an atom is a valid s-expression (by the inductive definition of s-expressions). You should not require to start with a `(`.